<h1>nf-core/nfmitnanext pipeline parameters</h1>
<p>mitochondrial nanopore reads extractor</p>
<h2>Input/output options</h2>
<p>Define where the pipeline should find input data and save output data.</p>
<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Description</th>
      <th>Type</th>
      <th>Default</th>
      <th>Required</th>
      <th>Hidden</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>input</code></td>
      <td>
        Path to comma-separated file containing information about the samples in the experiment.
        <details>
          <summary>Help</summary>
          <small
            >You will need to create a design file with information about the samples in your experiment before running
            the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 3
            columns, and a header row. See
            <a href="https://nf-co.re/nfmitnanext/usage#samplesheet-input">usage docs</a>.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td>True</td>
      <td></td>
    </tr>
    <tr>
      <td><code>outdir</code></td>
      <td>
        The output directory where the results will be saved. You have to use absolute paths to storage on Cloud
        infrastructure.
      </td>
      <td><code>string</code></td>
      <td></td>
      <td>True</td>
      <td></td>
    </tr>
    <tr>
      <td><code>email</code></td>
      <td>
        Email address for completion summary.
        <details>
          <summary>Help</summary>
          <small
            >Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when
            the workflow exits. If set in your user config file (<code>~/.nextflow/config</code>) then you don't need to
            specify this on the command line for every run.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
  </tbody>
</table>
<h2>Reference genome options</h2>
<p>Reference genome related files and options required for the workflow.</p>
<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Description</th>
      <th>Type</th>
      <th>Default</th>
      <th>Required</th>
      <th>Hidden</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>genome</code></td>
      <td>
        Name of iGenomes reference.
        <details>
          <summary>Help</summary>
          <small
            >If using a reference genome configured in the pipeline using iGenomes, use this parameter to give the ID
            for the reference. This is then used to build the full paths for all required reference genome files e.g.
            <code>--genome GRCh38</code>. <br /><br />See the
            <a href="https://nf-co.re/usage/reference_genomes">nf-core website docs</a> for more details.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>fasta</code></td>
      <td>
        Path to FASTA genome file.
        <details>
          <summary>Help</summary>
          <small
            >This parameter is <em>mandatory</em> if <code>--genome</code> is not specified. If you don't have a BWA
            index available this will be generated for you automatically. Combine with <code>--save_reference</code> to
            save BWA index for future runs.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>igenomes_ignore</code></td>
      <td>
        Do not load the iGenomes reference config.
        <details>
          <summary>Help</summary>
          <small
            >Do not load <code>igenomes.config</code> when running the pipeline. You may choose this option if you
            observe clashes between custom parameters and those supplied in <code>igenomes.config</code>.</small
          >
        </details>
      </td>
      <td><code>boolean</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>igenomes_base</code></td>
      <td>The base path to the igenomes reference files</td>
      <td><code>string</code></td>
      <td>s3://ngi-igenomes/igenomes/</td>
      <td></td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<h2>Institutional config options</h2>
<p>Parameters used to describe centralised config profiles. These should not be edited.</p>
<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Description</th>
      <th>Type</th>
      <th>Default</th>
      <th>Required</th>
      <th>Hidden</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>custom_config_version</code></td>
      <td>Git commit id for Institutional configs.</td>
      <td><code>string</code></td>
      <td>master</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>custom_config_base</code></td>
      <td>
        Base directory for Institutional configs.
        <details>
          <summary>Help</summary>
          <small
            >If you're running offline, Nextflow will not be able to fetch the institutional config files from the
            internet. If you don't need them, then this is not a problem. If you do need them, you should download the
            files from the repo and tell Nextflow where to find them with this parameter.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td>https://raw.githubusercontent.com/nf-core/configs/master</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>config_profile_name</code></td>
      <td>Institutional config name.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>config_profile_description</code></td>
      <td>Institutional config description.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>config_profile_contact</code></td>
      <td>Institutional config contact information.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>config_profile_url</code></td>
      <td>Institutional config URL link.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<h2>Generic options</h2>
<p>Less common options for the pipeline, typically set in a config file.</p>
<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Description</th>
      <th>Type</th>
      <th>Default</th>
      <th>Required</th>
      <th>Hidden</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>version</code></td>
      <td>Display version and exit.</td>
      <td><code>boolean</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>publish_dir_mode</code></td>
      <td>
        Method used to save pipeline results to output directory.
        <details>
          <summary>Help</summary>
          <small
            >The Nextflow <code>publishDir</code> option specifies which intermediate files should be saved to the
            output directory. This option tells the pipeline what method should be used to move these files. See
            <a href="https://www.nextflow.io/docs/latest/process.html#publishdir">Nextflow docs</a> for details.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td>copy</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>email_on_fail</code></td>
      <td>
        Email address for completion summary, only when pipeline fails.
        <details>
          <summary>Help</summary>
          <small
            >An email address to send a summary email to when the pipeline is completed - ONLY sent if the pipeline does
            not exit successfully.</small
          >
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>plaintext_email</code></td>
      <td>Send plain-text email instead of HTML.</td>
      <td><code>boolean</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>max_multiqc_email_size</code></td>
      <td>File size limit when attaching MultiQC reports to summary emails.</td>
      <td><code>string</code></td>
      <td>25.MB</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>monochrome_logs</code></td>
      <td>Do not use coloured log outputs.</td>
      <td><code>boolean</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>hook_url</code></td>
      <td>
        Incoming hook URL for messaging service
        <details>
          <summary>Help</summary>
          <small>Incoming hook URL for messaging service. Currently, MS Teams and Slack are supported.</small>
        </details>
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>multiqc_config</code></td>
      <td>Custom config file to supply to MultiQC.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>multiqc_logo</code></td>
      <td>Custom logo file to supply to MultiQC. File name must also be set in the MultiQC config file</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>multiqc_methods_description</code></td>
      <td>Custom MultiQC yaml file containing HTML including a methods description.</td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <td><code>validate_params</code></td>
      <td>Boolean whether to validate parameters against the schema at runtime</td>
      <td><code>boolean</code></td>
      <td>True</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>pipelines_testdata_base_path</code></td>
      <td>Base URL or local path to location of pipeline test dataset files</td>
      <td><code>string</code></td>
      <td>https://raw.githubusercontent.com/nf-core/test-datasets/</td>
      <td></td>
      <td>True</td>
    </tr>
    <tr>
      <td><code>trace_report_suffix</code></td>
      <td>
        Suffix to add to the trace report filename. Default is the date and time in the format yyyy-MM-dd_HH-mm-ss.
      </td>
      <td><code>string</code></td>
      <td></td>
      <td></td>
      <td>True</td>
    </tr>
  </tbody>
</table>
